### PR TITLE
Add further transform types

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
Implement Rotation, Scaling and Translation data types and operators to combine each of these with each other and the existing Affine3d/Matrix3d types.

Related to https://github.com/scipp/scipp/issues/2252

### CODE REVIEW

File: lib/core/include/scipp/core/element/arithmetic.h [hunk 1]
```diff
@@ -87,25 +88,72 @@ struct subtract_types_t {
...
+      std::tuple<std::tuple<scipp::core::ScalingTransform,
+                            scipp::core::ScalingTransform>>(),
+      std::tuple<std::tuple<scipp::core::ScalingTransform, Eigen::Matrix3d>>(),
+
+      std::tuple<std::tuple<Eigen::Matrix3d, Eigen::Affine3d>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, Eigen::Vector3d>>(),
+      std::tuple<
+          std::tuple<Eigen::Matrix3d, scipp::core::TranslationTransform>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, scipp::core::RotationTransform>>(),
+      std::tuple<std::tuple<Eigen::Matrix3d, scipp::core::ScalingTransform>>(),
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/element/arithmetic.h:138]: Except for the `Vector3d` cases, this looks like all possible permutations? Have a look at the helpers in `core/transform_common.h` to define these in a more concise manner. `pair_product_t` is probably doing what you need.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  Eigen::Quaterniond quat;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  RotationTransform() : quat(Eigen::Quaterniond::Identity()){};
+  explicit RotationTransform(const Eigen::Quaterniond &x) : quat(x){};
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const {
+    return quat.toRotationMatrix();
+  }
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:26]: Why is this needed?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+
+  inline bool operator==(const RotationTransform &other) const {
+    return this->matrix() == other.matrix();
+  }
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:30]: Is this implementation supposed to treat equivalent rotations as equal? I don't feel that that is a good default behavior for `operator==`? If we want to support checking for equivalence that should be done as a special explicit operation.

Here, I think we should just compare the coefficients.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+
+#include "scipp/core/dtype.h"
+
+namespace scipp::core {
+
+class RotationTransform {
+private:
+  // Store as quaterniond as this is more space efficient than storing as matrix
+  // (4 doubles for quat vs 9 doubles for 3x3 matrix).
+  Eigen::Quaterniond quat;
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:17]: We prefix members with `m_`

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+  ScalingTransform()
+      : mat(Eigen::Matrix3d::Identity().diagonal().asDiagonal()){};
+  explicit ScalingTransform(const Eigen::DiagonalMatrix<double, 3> &x)
+      : mat(x){};
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const { return mat; }
+
+  inline bool operator==(const ScalingTransform &other) const {
+    return this->matrix() == other.matrix();
+  }
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:48]: Seems inefficient? Just compare the diagonal values?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+};
+
+class TranslationTransform {
+private:
+  Eigen::Vector3d vec;
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:53]: Any reason why this is not using `Eigen::Translation`?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  TranslationTransform() : vec(Eigen::Vector3d(0, 0, 0)){};
+  explicit TranslationTransform(const Eigen::Vector3d &x) : vec(x){};
+
+  [[nodiscard]] Eigen::Vector3d vector() const { return vec; }
+
+  inline bool operator==(const TranslationTransform &other) const {
+    return this->vector() == other.vector();
+  }
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:64]: `inline` is redundant when defining a function inline in the class?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+};
+
+[[nodiscard]] inline RotationTransform operator*(const RotationTransform &lhs,
+                                                 const RotationTransform &rhs) {
+  return RotationTransform(Eigen::Quaterniond(lhs.matrix() * rhs.matrix()));
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:69]: Why is this converted to matrices and then back to a quaternion? Is there a problem with operating on the actual quaternions?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+};
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const RotationTransform &lhs,
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:72]: I feel we should drop the `Transform` suffix. Why not just `Rotation` and `Translation`? If in doubt, add a namespace for the transforms, i.e., make it `core::spatial::Rotation` or something like that?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                                               const ScalingTransform &rhs) {
+  return lhs.matrix() * rhs.matrix();
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:74]: Can't you let `Eigen` figure out the details, instead of working with matrices manually?
```suggestion
                                               const ScalingTransform &rhs) {
  return lhs.get() * rhs.get();
```
where the `get()` method returns the underlying `Eigen` object.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const Eigen::Matrix3d &lhs,
+                                               const RotationTransform &rhs) {
+  return lhs * rhs.matrix();
+}
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const Eigen::Matrix3d &lhs,
+                                               const ScalingTransform &rhs) {
+  return lhs * rhs.matrix();
+}
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:186]: There is a ton of duplication above. Can you avoid it by using a template (or a small number of templates)?

File: lib/variable/include/scipp/variable/variable.tcc [hunk 12]
```diff
@@ -38,12 +38,21 @@ template <class T> auto &cast(Variable &var) {
...
+                                      scipp::core::ScalingTransform> ||
+                       std::is_same_v<std::decay_t<T>,
+                                      scipp::core::RotationTransform>) {
+    return t.matrix().operator()(I);
+  } else if constexpr (std::is_same_v<std::decay_t<T>,
+                                      scipp::core::TranslationTransform>) {
+    return t.vector().operator()(I);
+  } else if constexpr (core::has_eval_v<std::decay_t<T>>) {
+    return t.operator()(I);
+  } else {
```
Comment (SimonHeybrock) [inline lib/variable/include/scipp/variable/variable.tcc:53]: At least some of the `matrix()` methods return by value, so I feel bad things will/may happen here. I think this can probably be avoided entirely, just expose access to the actual components, not to the equivalent matrices or vectors.

File: lib/variable/variable_instantiate_linalg.cpp [hunk 13]
```diff
@@ -58,6 +59,10 @@ std::vector<std::string> element_keys(const Variable &var) {
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(vector_3_float64, Eigen::Vector3d, double)
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(matrix_3_float64, Eigen::Matrix3d, double)
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(affine_transform, Eigen::Affine3d, double)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(rotation, scipp::core::RotationTransform)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(scaling, scipp::core::ScalingTransform)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(translation,
+                                   scipp::core::TranslationTransform)
```
Comment (SimonHeybrock) [inline lib/variable/variable_instantiate_linalg.cpp:65]: As mentioned on the call, I believe we must use structured variables here.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 6 above)
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:53]: The [eigen docs](https://eigen.tuxfamily.org/dox/classEigen_1_1Translation.html) specifically recommend against using `Eigen::Translation` to *store* a translation. To quote from the docs:

> This class is not aimed to be used to store a translation transformation, but rather to make easier the constructions and updates of Transform objects.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+
+template<class T> struct asEigenType_t { using type = T; };
+template<> struct asEigenType_t<Rotation> { using type = Eigen::Quaterniond; };
+template<> struct asEigenType_t<Scaling> { using type = Eigen::Matrix3d; };
+template<> struct asEigenType_t<Translation> { using type = Eigen::Translation<double, 3>; };
+
+template<typename T, typename R = asEigenType_t<T>::type> inline const R asEigenType(const T &obj) { return obj; };
+template<> inline const Eigen::Quaterniond asEigenType(const Rotation &obj) { return obj.quat(); }
+template<> inline const Eigen::Matrix3d asEigenType(const Scaling &obj) { return obj.matrix(); }
+template<> inline const Eigen::Translation<double, 3> asEigenType(const Translation &obj) { return Eigen::Translation<double, 3>(obj.vector()); }
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:102]: Should probably not return by value in any of the above?

File: lib/variable/variable_instantiate_linalg.cpp [hunk 15]
```diff
@@ -58,6 +59,10 @@ std::vector<std::string> element_keys(const Variable &var) {
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(vector_3_float64, Eigen::Vector3d, double)
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(matrix_3_float64, Eigen::Matrix3d, double)
 INSTANTIATE_STRUCTURE_ARRAY_VARIABLE(affine_transform, Eigen::Affine3d, double)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(rotation, scipp::core::Rotation)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(scaling, scipp::core::Scaling)
+INSTANTIATE_ELEMENT_ARRAY_VARIABLE(translation,
+                                   scipp::core::Translation)
```
Comment (SimonHeybrock) [inline lib/variable/variable_instantiate_linalg.cpp:65]: These should be structured variables.

**Review by SimonHeybrock — CHANGES_REQUESTED** [review]

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+          return m_quat.w();
+      else if (index == 1)
+          return m_quat.x();
+      else if (index == 2)
+          return m_quat.y();
+      else if (index == 3)
+          return m_quat.z();
+      else
+          throw std::out_of_range("invalid index in get for Rotation");
+  }
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:49]: Fixed the order. 

It's required by `get` in `variable.tcc`. You mentioned in a previous comment that we should just expose the individual coefficients to that function, not the whole eigen object.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 17]
```diff
@@ -0,0 +1,174 @@
...
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const { return m_mat; }
+
+  bool operator==(const Scaling &other) const {
+    return m_mat.diagonal() == other.m_mat.diagonal();
+  }
+
+  const double get(const int index) const {
+      return m_mat.diagonal()(index);
+  }
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:71]: As above, used by `get` in `variable.tcc`

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 18]
```diff
@@ -0,0 +1,174 @@
...
+
+  [[nodiscard]] Eigen::Vector3d vector() const { return m_vec; }
+
+  bool operator==(const Translation &other) const {
+    return m_vec == other.m_vec;
+  }
+
+  const double get(const int index) const {
+      return m_vec(index);
+  }
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:91]: As above, used by `get` in `variable.tcc`

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 16 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:49]: Are you sure it is getting used? Because:

- In `variable.tcc` we require access with a compile-time index, not run-time as in your implementation.
- The `get()` above returns by value, but in `variable.tcc` we need to get the address of the underlying element, i.e., we need a reference.

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:71]: See above, as it is this cannot work.

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 18 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:91]: See above, this cannot work as it is implemented now.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 19]
```diff
@@ -0,0 +1,174 @@
...
+[[nodiscard]] inline Rotation operator*(const Rotation &lhs,
+                                                 const Rotation &rhs) {
+  return Rotation(lhs.quat() * rhs.quat());
+};
+
+[[nodiscard]] inline Scaling operator*(const Scaling &lhs,
+                                                const Scaling &rhs) {
+  return Scaling(
+      (lhs.matrix() * rhs.matrix()).diagonal().asDiagonal());
+};
```
[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:161]: It doesn't look like it to me - here is the error I get in this function if I try passing through `Scaling` as an `Eigen::DiagonalMatrix<...>` (as opposed to `Eigen::Matrix3d`) everywhere, and change this function to simply be `return Scaling(lhs.matrix() * rhs.matrix());`:

```
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): error C2678: binary '*': no operator found which takes a left-hand operand of type 'Eigen::DiagonalMatrix<double,3,3>' (or there is no acceptable conversion) (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(76,5): message : could be 'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::Matrix<double,3,1,0,3,1>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &) const' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(157,30): message : or       'scipp::core::Scaling scipp::core::operator *(const scipp::core::Scaling &,const scipp::core::Scaling &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(152,31): message : or       'scipp::core::Rotation scipp::core::operator *(const scipp::core::Rotation &,const scipp::core::Rotation &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(82,5): message : or       'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>,const Eigen::Matrix<double,3,1,0,3,1>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &,const Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>> &)' [found using argument-dependent lookup] (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): message : while trying to match the argument list '(Eigen::DiagonalMatrix<double,3,3>, Eigen::DiagonalMatrix<double,3,3>)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\dtype.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): error C2678: binary '*': no operator found which takes a left-hand operand of type 'Eigen::DiagonalMatrix<double,3,3>' (or there is no acceptable conversion) (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(76,5): message : could be 'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::Matrix<double,3,1,0,3,1>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &) const' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(157,30): message : or       'scipp::core::Scaling scipp::core::operator *(const scipp::core::Scaling &,const scipp::core::Scaling &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(152,31): message : or       'scipp::core::Rotation scipp::core::operator *(const scipp::core::Rotation &,const scipp::core::Rotation &)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\.conan\data\eigen\3.3.9\_\_\package\5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9\include\eigen3\Eigen\src/Core/DiagonalMatrix.h(82,5): message : or       'const Eigen::DiagonalWrapper<const Eigen::CwiseBinaryOp<Eigen::internal::scalar_product_op<double,double>,const Eigen::CwiseNullaryOp<Eigen::internal::scalar_constant_op<double>,const Eigen::Matrix<double,3,1,0,3,1>>,const Eigen::Matrix<double,3,1,0,3,1>>> Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>>::operator *(const double &,const Eigen::DiagonalBase<Eigen::DiagonalMatrix<double,3,3>> &)' [found using argument-dependent lookup] (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\include\scipp/core/spatial_transforms.h(159,45): message : while trying to match the argument list '(Eigen::DiagonalMatrix<double,3,3>, Eigen::DiagonalMatrix<double,3,3>)' (compiling source file C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\lib\core\subbin_sizes.cpp) [C:\Users\wilt\Documents\11015_ESS_SCIPP\scipp\build\lib\core\scipp-core.vcxproj]
```

I can get around that by later adding a bunch of `static_cast`s back to `Matrix3d` in various places but that does not feel like the right approach at all.

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 14 above)
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:102]: I think at least `Translation` needs to return by value, as it is constructing an `Eigen::Translation` object which would be out of scope if we return it by reference? This seems to be what other people online suggest anyway with eigen types - return by value and let the compiler do any return-value optimisations needed to avoid copies etc. The [eigen docs](https://eigen.tuxfamily.org/dox/group__TopicPassingByValue.html) are also fairly explicit that `...there is no problem with functions that return objects by value. `

Having said that, I did try to get the other template specializations to return by ref *anyway*, but couldn't quite figure out the right syntax for it... if you feel strongly about this could I have a hand with it please?

Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:102]: > I think at least Translation needs to return by value, as it is constructing an Eigen::Translation object which would be out of scope if we return it by reference?

Yes, since the translation is not stored in the wrapper. I suppose `Translation` does not actually copy the values?

> The eigen docs are also fairly explicit that ...there is no problem with functions that return objects by value. 

There is no *problem* with it, but that does not mean that it is optimal for performance.

> Having said that, I did try to get the other template specializations to return by ref anyway, but couldn't quite figure out the right syntax for it... if you feel strongly about this could I have a hand with it please?

I think we have to do this properly, yes. If you return by value in the wrong place, you end up with expression templates referencing things on the stack that are not in scope any more.

File: lib/core/include/scipp/core/spatial_transforms.h (hunk 19 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:161]: Then maybe we should just multiply the underlying diagonal terms?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:161]: Maybe using Eigen internally is pointless in that case, and we should consider implementing our `Scaling` class directly?

[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:161]: That would work for this specific function, but after this one if fixed (either by the approach you suggest or by explicit casts) then we run into similar problems in the templated `operator*` functions above, and it's not so clear what we would do there...

The problem is a little bit different because `Quat * DiagonalMatrix` in eigen is *defined*, but gives `Affine3d` rather than `Matrix3d`, but we need it to give `Matrix3d` because rotations/scaling/matrix won't have a unit associated (whereas translation and affine will). This can probably be made to work with enough template special cases and so on, but it seems much clearer to pass through the scaling as a matrix which makes eigen do the right thing in the first place?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:162]: Maybe we should consider dropping the separate "scaling" dtype altogether, and just use a matrix (or a double, if uniform) for this from the start?

[the remark this answers is not in the document]
Comment (Tom-Willemsen) [reply lib/core/include/scipp/core/spatial_transforms.h:162]: We could do that, but it means 3x the memory use compared to storing it as a diagonalmatrix and only converting it to matrix as needed? Perhaps we aren't worried about that?

I guess from the user API perspective we would still have some constructors for scaling (to be consistent with translation/rotation), but they would just return `dtype=matrix` as opposed to `dtype=scaling`?

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply lib/core/include/scipp/core/spatial_transforms.h:162]: > We could do that, but it means 3x the memory use compared to storing it as a diagonalmatrix and only converting it to matrix as needed? Perhaps we aren't worried about that?

I think this is something we can worry about when it turns out to be a problem? There may be very limited use of this dtype, so if it avoids some mess for now that would be the better choice. It is always easier to add something later than to remove it.

> I guess from the user API perspective we would still have some constructors for scaling (to be consistent with translation/rotation), but they would just return dtype=matrix as opposed to dtype=scaling?

Yes, I agree with that suggestion, similar to how we have `from_rotvec`.

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+
+[[nodiscard]] inline Translation operator*(const Translation &lhs,
+                                           const Translation &rhs) {
+  return Translation(lhs.vector() + rhs.vector());
+};
+
+template <> inline constexpr DType dtype<Eigen::Matrix3d>{5001};
+template <> inline constexpr DType dtype<Eigen::Affine3d>{5002};
+template <> inline constexpr DType dtype<Translation>{5003};
+template <> inline constexpr DType dtype<Quaternion>{5004};
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:137]: Confusing that this was split up in this way. Just put all the spatial transforms and eigen types starting at 4000?

File: lib/core/include/scipp/core/spatial_transforms.h [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+template <class T_LHS> struct applies_to_vector : std::false_type {};
+template <> struct applies_to_vector<Quaternion> : std::true_type {};
+template <> struct applies_to_vector<Translation> : std::true_type {};
+
+template <typename T_LHS>
+[[nodiscard]] inline std::enable_if_t<applies_to_vector<T_LHS>::value,
+                                      Eigen::Vector3d>
+operator*(const T_LHS &lhs, const Eigen::Vector3d &rhs) {
+  return asEigenType(lhs) * rhs;
+}
```
Comment (SimonHeybrock) [inline lib/core/include/scipp/core/spatial_transforms.h:122]: In this case there appear to be only 2 possible template instantiations. Wouldn't it be more readable to write these functions explicitly? Maybe even as a member of the respective classes, to help the compiler?

File: lib/core/test/spatial_transforms_test.cpp [hunk 22]
```diff
@@ -0,0 +1,98 @@
...
+#include "test_macros.h"
+
+using namespace scipp;
+
+static const double TOLERANCE = pow(10, -10);
+
+TEST(SpatialTransformsTest, combine_to_linear) {
+  Eigen::Quaterniond x1;
+  x1 = Eigen::AngleAxisd(pi<double>, Eigen::Vector3d::UnitX());
+  Eigen::Matrix3d x2(Eigen::Matrix3d::Identity());
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:18]: Multiplication with identity does not really test much?

File: lib/core/test/spatial_transforms_test.cpp [hunk 23]
```diff
@@ -0,0 +1,98 @@
...
+
+  Eigen::Matrix3d result = scipp::core::Quaternion(x1) * x2;
+  Eigen::Matrix3d expected(Eigen::Quaterniond(
+      Eigen::AngleAxisd(pi<double>, Eigen::Vector3d::UnitX())));
+
+  for (int row = 0; row < 3; ++row) {
+    for (int col = 0; col < 3; ++col) {
+      ASSERT_NEAR(result(row, col), expected(row, col), TOLERANCE);
+    }
+  }
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:28]: I vaguely remember that `Eigen` has functions/methods for approximate comparisons. Can you have a look?

File: lib/core/test/spatial_transforms_test.cpp [hunk 24]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+TEST(SpatialTransformsTest, combine_to_affine) {
+  Eigen::Quaterniond x1;
+  x1 = Eigen::AngleAxisd(pi<double>, Eigen::Vector3d::UnitX());
+  Eigen::Affine3d x2(Eigen::Affine3d::Identity());
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:34]: See above.

File: lib/core/test/spatial_transforms_test.cpp [hunk 25]
```diff
@@ -0,0 +1,98 @@
...
+  for (int row = 0; row < 4; ++row) {
+    for (int col = 0; col < 4; ++col) {
+      ASSERT_NEAR(result(row, col), expected(row, col), TOLERANCE);
+    }
+  }
+}
+
+TEST(SpatialTransformsTest, apply_rotation_to_vector) {
+  Eigen::Quaterniond x1;
+  x1 = Eigen::AngleAxisd(pi<double>, Eigen::Vector3d::UnitX());
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:49]: Similar to identity rotations above: Rotation by `pi` is a "special case" and may hide a variety of bugs.

File: lib/core/test/spatial_transforms_test.cpp [hunk 26]
```diff
@@ -0,0 +1,98 @@
...
+TEST(SpatialTransformsTest, apply_translation_to_vector) {
+  scipp::core::Translation x1(Eigen::Vector3d(4, 5, 6));
+  Eigen::Vector3d x2(Eigen::Vector3d(1, 2, 3));
+
+  Eigen::Vector3d result = x1 * x2;
+  Eigen::Vector3d expected(5, 7, 9);
+
+  for (int i = 0; i < 3; ++i) {
+    ASSERT_NEAR(result(i), expected(i), TOLERANCE);
+  }
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:69]: Shouldn't `operator==` work here?

File: lib/core/test/spatial_transforms_test.cpp [hunk 27]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+TEST(SpatialTransformsTest, combine_rotations) {
+  Eigen::Quaterniond x1;
+  x1 = Eigen::AngleAxisd(pi<double> / 2.0, Eigen::Vector3d::UnitX());
+  Eigen::Quaterniond x2;
+  x2 = Eigen::AngleAxisd(pi<double> / 2.0, Eigen::Vector3d::UnitX());
```
Comment (SimonHeybrock) [inline lib/core/test/spatial_transforms_test.cpp:88]: Also test with rotations around distinct axes. If we only test for same axis we cannot be sure that we got the order of rotations correct.

**Review by SimonHeybrock — APPROVED** [review]
Looks good now 👍
